### ISSUE TITLE
How to generalise `load_flan_ul2` to other models

### ISSUE THREAD

**Issue by AndreaSottana** [akkikiki/huggingface_examples#1]
Hello,

Thanks so much for your very helpful example of how to load `google/flan-ul2.` I am getting a similar issue when trying to load `facebook/opt-66b`. If I load it with a simple 
```python
from transformers import AutoTokenizer, AutoModelForCausalLM
tokenizer = AutoTokenizer.from_pretrained("facebook/opt-66b")
model = AutoModelForCausalLM.from_pretrained("facebook/opt-66b", device_map="auto")
```

when I run inference I get the error `RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!` which if I understand correctly means that layers with a residual connection have been split across multiple devices. In [your example](https://github.com/akkikiki/huggingface_examples/blob/main/examples/load_flan_ul2.py) for `google/flan-ul2` you use the `no_split_module_classes=["T5Block"]` arguments (line 14) to get around the issue, however this is very model specific.

Your line 
```python
device_map['lm_head'] = device_map["decoder.embed_tokens"]
```
also seems model specific.

Do you have any general tips for how to quickly find out what we should put in `no_split_module_classes=` for other models, e.g. `facebook/opt-66b`? 

In your specific example, how did you find that the layer class name was `T5Block`? If I run `print(model.state_dict().keys())` I do not see any layer explicitly called `T5Block` so I'm curious to understand your thinking process

**Comment by akkikiki** [comment]
Yeah, these residual layers are pretty model-specific.
Here's an example of loading opt-66b which worked on my environment  https://github.com/akkikiki/huggingface_examples/blob/main/examples/load_opt.py

**Comment by AndreaSottana** [comment]
Thanks a lot for confirming, and for the addition:+1:

**Comment by AndreaSottana** [comment]
I have one follow up. What is the actual effect of using `load_in_8bit=True`? Does it slightly lower model's accuracy in the same way mixed precision does (`dtype=torch.float16`) or is it not at all related?

**Comment by akkikiki** [comment]
Yeah, it's basically another way of doing mixed precision inference. Take a look at https://huggingface.co/blog/hf-bitsandbytes-integration
